### PR TITLE
carved out hearbeat component from Node struct

### PR DESCRIPTION
**Pull request by iulianpascalau**
- carved out heartbeat component from node
- the heartbeat can no longer be disabled from config

### CODE REVIEW

File: heartbeat/componentHandler/heartbeatHandler.go [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+	AntifloodHandler         heartbeat.P2PAntifloodHandler
+	PeerBlackListHandler     heartbeat.BlackListHandler
+	ValidatorPubkeyConverter state.PubkeyConverter
+	EpochStartTrigger        sharding.EpochHandler
+	EpochStartRegistration   sharding.EpochStartEventNotifier
+	Timer                    heartbeat.Timer
+	GenesisTime              time.Time
+	VersionNumber            string
+	PeerShardMapper          heartbeat.NetworkShardingCollector
+	SizeCheckDelta           uint32
```
Comment (sasurobert) [inline heartbeat/componentHandler/heartbeatHandler.go:49]: do we still need this ?

File: heartbeat/componentHandler/heartbeatHandler.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+func (hbh *HeartbeatHandler) startSendingHeartbeats() {
+	r := rand.New(rand.NewSource(time.Now().Unix()))
+	cfg := hbh.arg.HeartbeatConfig
+
+	log.Debug("heartbeat's endless sending go routine started")
+
+	for {
+		diffSeconds := cfg.MaxTimeToWaitBetweenBroadcastsInSec - cfg.MinTimeToWaitBetweenBroadcastsInSec
+		diffNanos := int64(diffSeconds) * time.Second.Nanoseconds()
```
Comment (sasurobert) [inline heartbeat/componentHandler/heartbeatHandler.go:221]: do you really need this precision with nanoseconds ?

File: heartbeat/componentHandler/heartbeatHandler.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+	validators, err := hbh.arg.ValidatorStatistics.GetValidatorInfoForRootHash(latestHash)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return validators, nil, nil
+}
+
+func (hbh *HeartbeatHandler) startSendingHeartbeats() {
```
Comment (sasurobert) [inline heartbeat/componentHandler/heartbeatHandler.go:213]: rename to sendingHeartBeats ?

File: heartbeat/componentHandler/heartbeatHandler.go [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+	return hbh.monitor
+}
+
+// Sender returns the sender component
+func (hbh *HeartbeatHandler) Sender() *process.Sender {
+	return hbh.sender
+}
+
+// Close will close the endless running go routine
+func (hbh *HeartbeatHandler) Close() error {
```
Comment (sasurobert) [inline heartbeat/componentHandler/heartbeatHandler.go:270]: IsInterfaceNil

File: heartbeat/interface.go [hunk 5]
```diff
@@ -37,8 +50,8 @@ type Timer interface {
 type HeartbeatStorageHandler interface {
 	LoadGenesisTime() (time.Time, error)
 	UpdateGenesisTime(genesisTime time.Time) error
-	LoadHbmiDTO(pubKey string) (*HeartbeatDTO, error)
-	SavePubkeyData(pubkey []byte, heartbeat *HeartbeatDTO) error
+	LoadHbmiDTO(pubKey string) (*data.HeartbeatDTO, error)
```
Comment (sasurobert) [inline heartbeat/interface.go:53]: LoadHeartBeatDTO

File: heartbeat/mock/networkShardingCollectorStub.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
+package mock
+
+import "github.com/ElrondNetwork/elrond-go/p2p"
+
+type NetworkShardingCollectorStub struct {
+	UpdatePeerIdPublicKeyCalled  func(pid p2p.PeerID, pk []byte)
+	UpdatePublicKeyShardIdCalled func(pk []byte, shardId uint32)
+	UpdatePeerIdShardIdCalled    func(pid p2p.PeerID, shardId uint32)
+}
+
```
Comment (sasurobert) [inline heartbeat/mock/networkShardingCollectorStub.go:10]: missing comments

File: integrationTests/node/heartbeat/heartbeat_test.go [hunk 7]
```diff
@@ -12,9 +12,11 @@ import (
 	"github.com/ElrondNetwork/elrond-go/crypto/signing"
 	"github.com/ElrondNetwork/elrond-go/crypto/signing/mcl"
 	mclsig "github.com/ElrondNetwork/elrond-go/crypto/signing/mcl/singlesig"
+	"github.com/ElrondNetwork/elrond-go/heartbeat/data"
+	mock2 "github.com/ElrondNetwork/elrond-go/heartbeat/mock"
```
Comment (sasurobert) [inline integrationTests/node/heartbeat/heartbeat_test.go:16]: rename mock 2 - should use only mocks from integration tests

File: node/options_test.go [hunk 8]
```diff
@@ -7,6 +7,7 @@ import (
 
 	"github.com/ElrondNetwork/elrond-go/data/blockchain"
 	"github.com/ElrondNetwork/elrond-go/data/endProcess"
+	mock2 "github.com/ElrondNetwork/elrond-go/heartbeat/mock"
```
Comment (sasurobert) [inline node/options_test.go:10]: use mock from node package

File: integrationTests/mock/heartbeatStorerStub.go [hunk 9]
```diff
@@ -3,15 +3,15 @@ package mock
...
+	"github.com/ElrondNetwork/elrond-go/heartbeat/data"
 )
 
 // HeartbeatStorerStub -
 type HeartbeatStorerStub struct {
 	LoadGenesisTimeCalled   func() (time.Time, error)
 	UpdateGenesisTimeCalled func(genesisTime time.Time) error
-	LoadHbmiDTOCalled       func(pubKey string) (*heartbeat.HeartbeatDTO, error)
-	SavePubkeyDataCalled    func(pubkey []byte, heartbeat *heartbeat.HeartbeatDTO) error
+	LoadHbmiDTOCalled       func(pubKey string) (*data.HeartbeatDTO, error)
```
Comment (SebastianMarian) [inline integrationTests/mock/heartbeatStorerStub.go:13]: I think Hbmi could be replaced with Heartbeat everywhere in all these files

File: node/node.go [hunk 10]
```diff
@@ -928,168 +928,43 @@ func (n *Node) GetAccount(address string) (state.UserAccountHandler, error) {
...
 }
 
 // GetHeartbeats returns the heartbeat status for each public key defined in genesis.json
-func (n *Node) GetHeartbeats() []heartbeat.PubKeyHeartbeat {
-	if n.heartbeatMonitor == nil {
-		return nil
-	}
-	return n.heartbeatMonitor.GetHeartbeats()
+func (n *Node) GetHeartbeats() []heartbeatData.PubKeyHeartbeat {
+	return n.heartbeatHandler.Monitor().GetHeartbeats()
```
Comment (SebastianMarian) [inline node/node.go:967]: Check for nil for n.heartbeatHandler ?

**Comment by elrondscammer3** [comment]
Crooked sons of bitches. Give us our money back.
 BTC dump ERD dump BTC pump ERD dump BTC stable ERD dump. Don'T buy ER

File: heartbeat/interface.go (hunk 5 above)
Comment (elronscam8) [reply heartbeat/interface.go:53]: Beniamin Mincu 
ERD/Elrond Scam Coin. Team ERD is a con man. Don't buy ERD. I send this
 message every day to everyone in the group. Prevent more people from being conned by sharing with you.
  ERD BIG DUMP SELL NOW ...

File: integrationTests/node/heartbeat/heartbeat_test.go (hunk 7 above)
Comment (elronscam8) [reply integrationTests/node/heartbeat/heartbeat_test.go:16]: Crooked sons of bitches. Give us our money back.
 BTC dump ERD dump BTC pump ERD dump BTC stable ERD dump. Don'T buy ERD

File: integrationTests/mock/heartbeatStorerStub.go (hunk 9 above)
Comment (elronscam8) [reply integrationTests/mock/heartbeatStorerStub.go:13]: That's enough!! Cheating bastards.
The price keeps falling. People don't trust you. ERD doesn't buy. What do you boys do for a living?

**Comment by elronscam8** [comment]
That's enough!! Cheating bastards.
The price keeps falling. People don't trust you. ERD doesn't buy. What do you boys do for a living?

File: heartbeat/componentHandler/heartbeatHandler.go [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+}
+
+func (hbh *HeartbeatHandler) startSendingHeartbeats(ctx context.Context) {
+	r := rand.New(rand.NewSource(time.Now().Unix()))
+	cfg := hbh.arg.HeartbeatConfig
+
+	log.Debug("heartbeat's endless sending go routine started")
+
+	for {
+		diffSeconds := cfg.MaxTimeToWaitBetweenBroadcastsInSec - cfg.MinTimeToWaitBetweenBroadcastsInSec
```
Comment (AdoAdoAdo) [inline heartbeat/componentHandler/heartbeatHandler.go:220]: no need to compute timeToWait in every iteration, as it should always be the same, can be done outside of the loop

File: heartbeat/componentHandler/heartbeatHandler.go [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+
+		err := hbh.sender.SendHeartbeat()
+		if err != nil {
+			log.Debug("SendHeartbeat", "error", err.Error())
+		}
+	}
+}
+
+func (hbh *HeartbeatHandler) checkConfigParams(config config.HeartbeatConfig) error {
+	if config.DurationInSecToConsiderUnresponsive < 1 {
```
Comment (AdoAdoAdo) [inline heartbeat/componentHandler/heartbeatHandler.go:240]: maybe rename to be consistent with the other? ....InSec

File: heartbeat/componentHandler/heartbeatHandler.go (hunk 1 above)
Comment (iulianpascalau) [reply heartbeat/componentHandler/heartbeatHandler.go:49]: we need testing, added todo

File: heartbeat/componentHandler/heartbeatHandler.go (hunk 3 above)
Comment (iulianpascalau) [reply heartbeat/componentHandler/heartbeatHandler.go]: it was easier to implement this as the difference between 2 timestamps is expressed in nanoseconds.

File: heartbeat/componentHandler/heartbeatHandler.go (hunk 4 above)
Comment (iulianpascalau) [reply heartbeat/componentHandler/heartbeatHandler.go:270]: done

File: heartbeat/interface.go (hunk 5 above)
Comment (iulianpascalau) [reply heartbeat/interface.go:53]: done

File: heartbeat/mock/networkShardingCollectorStub.go (hunk 6 above)
Comment (iulianpascalau) [reply heartbeat/mock/networkShardingCollectorStub.go:10]: added comments

File: integrationTests/node/heartbeat/heartbeat_test.go (hunk 7 above)
Comment (iulianpascalau) [reply integrationTests/node/heartbeat/heartbeat_test.go:16]: done

File: node/options_test.go (hunk 8 above)
Comment (iulianpascalau) [reply node/options_test.go:10]: done

File: heartbeat/componentHandler/heartbeatHandler.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply heartbeat/componentHandler/heartbeatHandler.go:220]: done

File: heartbeat/mock/singlesignerMock.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+// IsInterfaceNil returns true if there is no value under the interface
+func (s *SinglesignMock) IsInterfaceNil() bool {
+	return s == nil
+}
+
+// SinglesignFailMock -
+type SinglesignFailMock struct {
+}
+
+// Sign Signs a message with using a single signature schnorr scheme
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply heartbeat/mock/singlesignerMock.go:38]: simplified comments

File: heartbeat/mock/singlesignerMock.go [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+	return s == nil
+}
+
+// SinglesignStub -
+type SinglesignStub struct {
+	SignCalled   func(private crypto.PrivateKey, msg []byte) ([]byte, error)
+	VerifyCalled func(public crypto.PublicKey, msg []byte, sig []byte) error
+}
+
+// Sign Signs a message with using a single signature schnorr scheme
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply heartbeat/mock/singlesignerMock.go:59]: simplified comments

File: integrationTests/mock/heartbeatStorerStub.go (hunk 9 above)
Comment (iulianpascalau) [reply integrationTests/mock/heartbeatStorerStub.go:13]: done

File: node/node.go (hunk 10 above)
Comment (iulianpascalau) [reply node/node.go:967]: added guards

File: heartbeat/errors.go [hunk 15]
```diff
@@ -89,3 +89,21 @@ var ErrZeroHbmiRefreshIntervalInSec = errors.New("zero hbmiRefreshInterval")
 // ErrZeroHideInactiveValidatorIntervalInSec signals that a zero value
 // was provided for the ErrZeroHideInactiveValidatorIntervalInSec
 var ErrZeroHideInactiveValidatorIntervalInSec = errors.New("zero hideInactiveValidatorIntervalInSec")
+
+// ErrNegativeDurationInSecToConsiderUnresponsive is raised when a value less than 1 has been provided
+var ErrNegativeDurationInSecToConsiderUnresponsive = errors.New("value DurationInSecToConsiderUnresponsive is less" +
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply heartbeat/errors.go:94]: changed to one liner (all three of them)

File: heartbeat/componentHandler/heartbeatHandler.go (hunk 11 above)
Comment (iulianpascalau) [reply heartbeat/componentHandler/heartbeatHandler.go:220]: agree, changed

File: heartbeat/componentHandler/heartbeatHandler.go (hunk 12 above)
Comment (iulianpascalau) [reply heartbeat/componentHandler/heartbeatHandler.go:240]: changed

File: heartbeat/componentHandler/heartbeatHandler.go [hunk 16]
```diff
@@ -216,9 +216,10 @@ func (hbh *HeartbeatHandler) startSendingHeartbeats(ctx context.Context) {
 
 	log.Debug("heartbeat's endless sending go routine started")
 
+	diffSeconds := cfg.MaxTimeToWaitBetweenBroadcastsInSec - cfg.MinTimeToWaitBetweenBroadcastsInSec
+	diffNanos := int64(diffSeconds) * time.Second.Nanoseconds()
+
 	for {
-		diffSeconds := cfg.MaxTimeToWaitBetweenBroadcastsInSec - cfg.MinTimeToWaitBetweenBroadcastsInSec
-		diffNanos := int64(diffSeconds) * time.Second.Nanoseconds()
 		randomNanos := r.Int63n(diffNanos)
```
Comment (AdoAdoAdo) [inline heartbeat/componentHandler/heartbeatHandler.go:223]: why not move these two line as well outside of the loop?

File: integrationTests/node/heartbeat/heartbeat_test.go [hunk 17]
```diff
@@ -12,9 +12,11 @@ import (
 	"github.com/ElrondNetwork/elrond-go/crypto/signing"
 	"github.com/ElrondNetwork/elrond-go/crypto/signing/mcl"
 	mclsig "github.com/ElrondNetwork/elrond-go/crypto/signing/mcl/singlesig"
+	"github.com/ElrondNetwork/elrond-go/heartbeat/data"
+	heartbeatMock "github.com/ElrondNetwork/elrond-go/heartbeat/mock"
```
Comment (sasurobert) [inline integrationTests/node/heartbeat/heartbeat_test.go:16]: only integrationsTests/mock - here you have heartbeat and node

**Pull request merged**